### PR TITLE
Save doc file node references and use in partial parsing

### LINKED ISSUE #3425: Save information about doc references for partial parsing
If a document file changes when the partial parse flag is on, we can't identify the nodes that use that document. Save the information about referenced docs in the node, so that when the document changes we can schedule those nodes for re-parsing.

### PR DESCRIPTION
**Pull request by gshank**
resolves #3425

### Description

When doc references are processed in parsing, save the unique_id of the node with the reference in the 'nodes' list in the doc source file. When the doc is changed and processed during partial parse, it will use the list of nodes to re-parse and update the docs.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/parser/manifest.py [hunk 1]
```diff
@@ -255,8 +255,8 @@ def load(self):
             # load up the Lookup objects to resolve them by name, so the SourceFiles store
             # the unique_id instead of the name. Sources are loaded from yaml files, so
             # aren't in place yet
-            self.manifest.ref_lookup
-            self.manifest.doc_lookup
+            self.manifest.rebuild_ref_lookup()
+            self.manifest.rebuild_doc_lookup
```
Comment (kwigley) [inline core/dbt/parser/manifest.py:259]:
```suggestion
            self.manifest.rebuild_doc_lookup()
```

**Review by kwigley — APPROVED** [review]
I didn't test this, but code changes look good to me.

**Review by jtcohen6 — APPROVED** [review]
This works for me in local testing. Nice work!

Intuitively, I was expecting an approach whereby docs blocks are included in the `depends_on` of each node. I don't have a strong reason to prefer it, though, and I'm very happy with this approach as well.

File: core/dbt/parser/manifest.py (hunk 1 above)
Comment (gshank) [reply core/dbt/parser/manifest.py:259]: Good catch.

**Comment by gshank** [comment]
I initially started using a docs list in the depends_on, but that involved a lot of changes (to tests particularly) and would be a schema change. Also we'd have to loop though ALL of the nodes, sources, docs, and macros looking for the depends_on.docs. Storing them in the docs file means that we can skip that.

**Comment by jtcohen6** [comment]
> Storing them in the docs file means that we can skip that.

Makes sense to me!

**Pull request merged**